### ISSUE TITLE
Deprecate TimeFieldSpec and start using only DateTimeFieldSpec

### ISSUE THREAD

**Issue by npawar** [linkedin/pinot#2756]
Currently, we can only have one time column in the table which is limiting some use cases . We should allow multiple time columns and even allow derived time columns. Derived columns can be useful in performing roll ups or generating star tree aggregate nodes. We need to have multiple time columns as opposed to storing all extra time columns in dimensions, because the time columns come with some special time specific properties like granularity  and format.
Brief steps to perform this migration:
1) Introduce DateTimeFieldSpec class, and List<DateTimeFieldSpec> in Schema - DONE
2) Replacing timeSpec with list of dateTimeSpec will mean we won't know which dateTimeSpec is the primary time column by looking at the schema alone. We need to ensure that in all places where primary column is needed, we have the table config available to know the time column name from validationConfig
3) 
i) Deprecate TIME, TimeFieldSpec, and all getters and setters of time from the schema. 
ii) Modify getDateTimeFieldSpec to return timeFieldSpec along with the dateTimeFieldSpecs.
iii) Modify code to stop using timeFieldSpec. Instead, read dateTimeFieldSpecs. 
iv) Deploy this code
4) Change Thirdeye and other internal applications. Replace their usages of timeSpec with the appropriate dateTimeFieldSpec equivalent
5) Delete timeSpec related methods and classes 

Will add more details to this and a wiki as I have more details figured out

**Comment by npawar** [comment]
Design Doc
https://docs.google.com/document/d/1SU1jCjfsIDSA960fD5YWQbD72p8UdGF0c7CroFNt9Ho/edit?usp=sharing

**Comment by snleee** [comment]
I want to discuss about enforcing a uniform type & format and only allow users to pick the granularity for the `primary time column`, which is configured in the table config.

- start with the primary time column (enforced to have `long, "1:milliseconds:epoch", user-defined-granularity`). 

- Users can initially try to use time conversion UDF for the time column. 

- If they need the better performance, they can add a new time column to `dateTimeFieldSpecs` with the specific time format & type that they want to materialize.

I think that having a uniform type & format for the primary column can simplify a lot of components because we won't need to support all different type formats when we need to play with the time. One example can benefit from this is that segment roll-up config can be very simple - getting only time granularity. 

Another huge benefit will be an easier integration with external UIs such as Superset. I believe that Pinot-Superset connector have some extra logic to convert the time column values into a format that Superset understands. Since pinot's time column format/type will be different for each table, then the external system now needs to know about Pinot's table configuration & schema (or separate config for each table on Superset) for correctly parsing the time column value.

Since we make the interface change here to the time column, it's going to be great if we can consider this option while doing this work. 

@kishoreg @npawar @mayankshriv @Jackie-Jiang  How do you guys think?

**Comment by npawar** [comment]
I agree this would be nice to have, and see the benefits we gain for merge/rollup and integrations.
I would prefer if we first get to making dateTimeFieldSpec the default, and then focus on this. Once we have only dateTimeFieldSpec, it'll be easier to add tooling and validations for achieving this. I feel this change is kinda tricky as it is, and don't want to mix up goals.
